### PR TITLE
Add "Spelling / Did you mean?" feature

### LINKED ISSUE #3816: Feature: Spelling / Did you mean?
**Is your feature request related to a problem? Please describe.**
At present, search engines such as Google will put a corrected / suggested correct spelling for a word at the top of search results if you happen to spell something incorrectly. 

<br />

**Describe the solution you'd like**

An attempt to automatically suggest a corrected word in the event that the user may not know the correct spelling for something.

<br />

**Describe alternatives you've considered**

No real alternative aside from just spelling the word correctly. I did review some different APIs for other languages and came across libraries such as https://serpapi.com/spell-check (just as an example).

https://apilayer.com/marketplace/dymt-api

<br />

**Additional context**

The obvious thing here is that it must be a way that continues to ensure privacy, but also doesn't railroad us with a bunch of restrictions / API limits, which greatly narrows down the options. I looked online to see if search engines such as Google offered this type of API for "Did you know", and I could not find anything other than Google offering a "Custom Search Engine" service, which still uses Google search, but narrows down what the user can search through.

k3GnzHtrgp

### PR DESCRIPTION
**Pull request by sepehr-rs**
## What does this PR do? 
This PR adds "Spelling / Did you mean?" feature, also adds spell check to preferences. 
     
 ## Why is this change important? 
 It will help users use searxng more conveniently and avoid spelling errors
      
## How to test this PR locally? 
Search in searxng with any spelling errors and it will detect it 
NOTE : This spelling functionality uses the `pyspellchecker` library, because this library only supports English, Spanish, French, Portuguese, German, Italian, Russian, Arabic, Basque, Latvian and Dutch, if any other languages are selected, the spell checker functionality will not run.  
## Related issues 
Closes #3816

### CODE REVIEW

**Comment by return42** [comment]
To fix code problems to pass black test use `make format.python` / see our [Development Quickstart](https://docs.searxng.org/dev/quickstart.html#development-quickstart) guide .. BTW: can you squash the 11 commits in just one / Thanks :+1:

**Comment by sepehr-rs** [comment]
> To fix code problems to pass black test use `make format.python` / see our [Development Quickstart](https://docs.searxng.org/dev/quickstart.html#development-quickstart) guide .. BTW: can you squash the 11 commits in just one / Thanks 👍

Thanks for your kind reply, I will do as you requested.

File: searx/templates/simple/preferences/spellcheck.html [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+        <option value="0"
+                {%- if spellcheck == '0' %} selected="selected" {%- endif -%}>
+                {{- _('Off') -}}
+        </option>{{- '' -}}
+      </select>{{- '' -}}
+    </div>{{- '' -}}
+    <div class="description">
+      {{- _('Spell check search queries') -}}
+    </div>{{- '' -}}
+  </fieldset>{{- '' -}}
```
Comment (Bnyro) [inline searx/templates/simple/preferences/spellcheck.html:18]: Since there are only two possible values (on/off), please make this a checkbox preference such as https://github.com/searxng/searxng/blob/master/searx/templates/simple/preferences/query_in_title.html.

File: searx/templates/simple/results.html [hunk 2]
```diff
@@ -92,6 +92,25 @@ <h4 id="corrections-title">{{ _('Try searching for:') }}</h4>
...
+      <input type="hidden" name="category_{{ category }}" value="1" >
+      {% endfor %}
+      <input type="hidden" name="pageno" value="{{ pageno+1 }}" >
+      <input type="hidden" name="language" value="{{ current_language }}" >
+      <input type="hidden" name="time_range" value="{{ time_range }}" >
+      <input type="hidden" name="safesearch" value="{{ safesearch }}" >
+      <input type="hidden" name="theme" value="{{ theme }}" >
+      {% if timeout_limit %}<input type="hidden" name="timeout_limit" value="{{ timeout_limit|e }}" >{% endif %}
+      {{- engine_data_form(engine_data) -}}
+      <span style="color: #ff7769; font-size: 18px"> Did you mean:
```
Comment (Bnyro) [inline searx/templates/simple/results.html:108]: Please don't hardcode style here - edit https://github.com/searxng/searxng/blob/master/searx/static/themes/simple/src/less/search.less with the CSS instead and run `static.build.commit` after.

Comment (Bnyro) [reply searx/templates/simple/results.html:108]: Apart from that I don't think we need a new color here, you can use an existing one of  https://github.com/searxng/searxng/blob/master/searx/static/themes/simple/src/less/definitions.less

**Review by Bnyro — CHANGES_REQUESTED** [review]
Additionally, please also add the `spellcheck` setting to `searx/settings.yml` and `searx/settings_defaults`.

**Comment by Bnyro** [comment]
The suggestions can be very funny sometimes, but I generally think that this is a really nice addition, great idea :)

I have left some comments, feel free to ask if something about my review is unclear :)

File: searx/webapp.py [hunk 3]
```diff
@@ -254,6 +255,17 @@ def code_highlighter(codelines, language=None):
     return html_code
 
 
+def spellcheck(query: str, user_language: str):
+    spell = SpellChecker(language=user_language)
+    spellcheck_list = query.split(' ')
+    spellcheck_list = [x for x in spellcheck_list if x != '']
+    has_misspelled_words = bool(spell.unknown(spellcheck_list))
```
Comment (Bnyro) [inline searx/webapp.py:264]:
```suggestion
    has_misspelled_words = any(spell.unknown(spellcheck_list))
```
should do the same I think and is more readable

File: searx/webapp.py [hunk 4]
```diff
@@ -254,6 +255,17 @@ def code_highlighter(codelines, language=None):
...
+    for word, i in zip(spellcheck_list, range(0, len(spellcheck_list))):
+        if word != '':
+            spellcheck_list[i] = spell.correction(word) if spell.correction(word) is not None else ''
```
Comment (Bnyro) [inline searx/webapp.py:267]:
```suggestion
            spellcheck_list[i] = spell.correction(word) or ''
```

**Comment by sepehr-rs** [comment]
Thanks @Bnyro for your kind reviews !, I will do as you requested,
Have a nice day :).

**Comment by sepehrrasooli** [comment]
Hi @Bnyro, I've been running in circles with this issue and I really don't understand why it's not working
I changed the search.less file in the static directory and added the following lines :
```css
input[type="spellcheck_query"],
#results button[type="spellcheck_query"],
.spellcheck_query {
  background: none;
  border: none;
  color: var(--color-search-background-hover);
  cursor: pointer;
  font-size: 20px;
}
```
And I looked into the [Development Quickstart ](https://docs.searxng.org/dev/quickstart.html#development-quickstart)guide and these was a mention of using 
```bash
make static.build.commit
```
And when i run that i get the following error :
```bash
...
  tests/unit/test_webutils.py
  utils/get_setting.py
  utils/lxc-searxng.env
  utils/makefile.include
  utils/makefile.lxc
  utils/searxng_check.py
  utils/templates/etc/httpd/sites-available/searxng.conf
  utils/templates/etc/httpd/sites-available/searxng.conf:socket
  utils/templates/etc/nginx/default.apps-available/searxng.conf
  utils/templates/etc/nginx/default.apps-available/searxng.conf:socket
  utils/templates/etc/searxng/settings.yml
  utils/templates/etc/uwsgi/apps-archlinux/searxng.ini
  utils/templates/etc/uwsgi/apps-archlinux/searxng.ini:socket
  utils/templates/etc/uwsgi/apps-available/searxng.ini
  utils/templates/etc/uwsgi/apps-available/searxng.ini:socket
  utils/templates/lib/systemd/system/filtron.service
  utils/templates/lib/systemd/system/morty.service
  utils/templates/lib/systemd/system/searxng-redis.service
make: *** [Makefile:94: static.build.commit] Error 1
```
And whatever I do, clearing cache and trying again and etc., the changes dont apply to the css files.
Could you help me please, I've done all I could :(.
Thanks in advance.

**Comment by sepehr-rs** [comment]
Hi @Bnyro, I was able to compile and apply the static files,I kept getting errors because of an stupid permission problem.
But i was not able to change the css attributes of the input tag, this line :
```html
      <input style="background: none; border: none; color: #1a0dab; cursor: pointer; font-size: 20px" type="submit" role="link" value="{{ spellcheck_query }}">
```
This is because of the `type='submit'` attribute, do you think hard coding the style is ok here ?

**Comment by return42** [comment]
FYI: I rebased to clean up commit history and solve a conflict with the master branch.

**Comment by sepehr-rs** [comment]
Any updates ?!
@Bnyro  @return42

**Comment by Bnyro** [comment]
> Any updates ?! @Bnyro @return42
>
Well, we're waiting for you to implement my review comments...

**Comment by Bnyro** [comment]
Apart from that we're not sure about the memory footprint of using that library, that must be tested intensively before considering to merge this PR.

**Comment by sepehr-rs** [comment]
> > Any updates ?! @Bnyro @return42
> 
> Well, we're waiting for you to implement my review comments...

I did, but I faced some problems that I couldn't solve without [hard coding the style attribute](https://github.com/searxng/searxng/pull/3837#issuecomment-2370259100), So I asked you for some suggestions but got no response 😅



> Apart from that we're not sure about the memory footprint of using that library, that must be tested intensively before considering to merge this PR.

Could you please elaborate on this more ? I want to know how I can help.

**Comment by return42** [comment]
>> Apart from that we're not sure about the memory footprint of using that library,
> Could you please elaborate on this more ? I want to know how I can help.

The database for languages is contained in the Python package, i.e. every walker process of the WEB application loads it into the memory (with at least 4 processes, this is multiplied by 4).  Depending on the size of the database and how this is implemented, this can take up a lot of memory.   

Another disadvantage of this solution is that the database is permanently bound to the distribution of the Python package, see https://github.com/barrust/pyspellchecker/tree/master/spellchecker/resources

It is therefore generally advisable to have applications such as “suggestions” as a separate service and to integrate the service into SearXNG.


Basically, we already have something that does exactly this .. it is the [suggestions](https://github.com/searxng/searxng/blob/master/searx/templates/simple/elements/suggestions.html) implementation in SearXNG.

**Comment by return42** [comment]
A very simple example for the "suggestions" is given in [example-plugin](https://docs.searxng.org/dev/plugins.html#example-plugin) but we use it most often to fetch the suggestions from the origin search engines. For example, we use the suggestions from the google engine:

https://github.com/searxng/searxng/blob/0245e82bd24fce9a0fe5194dcac9241e8d0efdc7/searx/engines/google.py#L386-L389

**Comment by sepehr-rs** [comment]
> > > Apart from that we're not sure about the memory footprint of using that library,
> > > Could you please elaborate on this more ? I want to know how I can help.
> 
> The database for languages is contained in the Python package, i.e. every walker process of the WEB application loads it into the memory (with at least 4 processes, this is multiplied by 4). Depending on the size of the database and how this is implemented, this can take up a lot of memory.
> 
> Another disadvantage of this solution is that the database is permanently bound to the distribution of the Python package, see https://github.com/barrust/pyspellchecker/tree/master/spellchecker/resources
> 
> It is therefore generally advisable to have applications such as “suggestions” as a separate service and to integrate the service into SearXNG.
> 
> Basically, we already have something that does exactly this .. it is the [suggestions](https://github.com/searxng/searxng/blob/master/searx/templates/simple/elements/suggestions.html) implementation in SearXNG.

Thank you for the complete explanation :)
I understand the problems it would cause but I don't understand how we could take advantage of the "suggestions" feature.
Should we query the suggestions from the origin search engine and then display the first suggestion to the user?
I would greatly appreciate your guidance on this.

**Comment by aikooo7** [comment]
Hey,

> Additionally, please also add the `spellcheck` setting to `searx/settings.yml` and `searx/settings_defaults`.

Is this the thing blocking this pr from being merged?

**Comment by sepehr-rs** [comment]
> Hey,
> 
> > Additionally, please also add the `spellcheck` setting to `searx/settings.yml` and `searx/settings_defaults`.
> 
> Is this the thing blocking this pr from being merged?

Hi, 
No, the thing blocking this from being merged is two issues that @Bnyro and @return42 mentioned and requested changes about them but I can't solve the issues without further guidance. The issues being :
1. Hard coding the style attribute, as said here : 

> Please don't hardcode style here - edit https://github.com/searxng/searxng/blob/master/searx/static/themes/simple/src/less/search.less with the CSS instead and run `static.build.commit` after.

2. Memory concerns and using the suggestions feature instead of the pyspellchecker library, as said here :

> A very simple example for the "suggestions" is given in [example-plugin](https://docs.searxng.org/dev/plugins.html#example-plugin) but we use it most often to fetch the suggestions from the origin search engines. For example, we use the suggestions from the google engine:
> 
> https://github.com/searxng/searxng/blob/0245e82bd24fce9a0fe5194dcac9241e8d0efdc7/searx/engines/google.py#L386-L389

**Comment by aikooo7** [comment]
This is a feature I really wanted, I will go ahead and try to fix this pr as it appears stalled

**Comment by aikooo7** [comment]
@Bnyro While trying to migrate this pr to the current master I am not able to make preferences work, looking at the repo I assume now the proper way is make a plugin like https://github.com/searxng/searxng/blob/master/searx/plugins/calculator.py ?

**Comment by Bnyro** [comment]
> @Bnyro While trying to migrate this pr to the current master I am not able to make preferences work, looking at the repo I assume now the proper way is make a plugin like https://github.com/searxng/searxng/blob/master/searx/plugins/calculator.py ?
>
Yes, the plugic ecosystem has been rewritten very recently so that plugins should now subclass `SXNGPlugin` as in the calculator plugin for example.

---

- https://github.com/searxng/searxng/pull/4538
- [Plugin Development](https://docs.searxng.org/dev/plugins/development.html#module-searx.plugins)

**Comment by aikooo7** [comment]
But for me to make preferences.html work again I need to rewrite this pr as
a plugin?

**Comment by Bnyro** [comment]
> But for me to make preferences.html work again I need to rewrite this pr as a plugin?
>
No, but I think this pr would generally be suited well for being made a plugin, I don't think it should be in SearXNG core due to the reasons discussed above.

**Comment by sepehr-rs** [comment]
Hey all, just wanted to quickly reflect on this PR since I noticed @aikooo7  has picked it up.

I really appreciate all the feedback and learned a lot in the process. Looking back, I think I’d be able to fix the CSS issues now with the experience I’ve gained over the past couple of months. That said, the memory concerns around the spellchecker library were beyond my scope at the time, especially as a beginner without the right perspective yet.

Thanks again for the opportunity, and thanks to @aikooo7 for stepping in to keep things moving forward. If there are other issues I can help with in the future, I’d be happy to contribute again.

P.S. If there’s anything I could’ve done better from your side, feel free to share. Always looking to improve.

**Comment by aikooo7** [comment]
My life is pretty busy right now so I can’t confirm if I will be able to implement this, I heard that [moa](https://github.com/moa-engine/MOA), a searxng fork has this feature, could we perhaps see how they done it and implement it here?

**Comment by return42** [comment]
Closed due to inactivity.
